blobdb migrator for compressing form XML data

snopoke opened the pull request.

snopoke committed:
blob migrator to compress form xml

snopoke committed:
make migrations functions

snopoke committed:
rename test

snopoke committed:
[ci skip] remove unused import

millerdev commented on corehq/blobs/migrate.py:
```diff
@@ -414,23 +451,26 @@ class BackendMigrator(Migrator):
+            iteration_key = f'{self.iteration_key}-{start}-{end}'
+        if domain:
+            self.reindexer.domain = domain
+            iteration_key = f'{self.iteration_key}-{domain}'
```
```suggestion
            iteration_key = f'{iteration_key}-{domain}'
```

millerdev replied:
I'm assuming the previous suggestion I left here was resolved by mistake because it looked so much like the other one. I think it still needs to be changed. If not, can you explain why not?
```suggestion
            iteration_key = f'{iteration_key}-{domain}'
```

millerdev commented on corehq/blobs/migrate.py:
```diff
@@ -414,23 +451,26 @@ class BackendMigrator(Migrator):
             (start, end) = date_range
             self.reindexer.date_range = date_range
-            iteration_key = '{}-{}-{}'.format(self.iteration_key, start, end)
+            iteration_key = f'{self.iteration_key}-{start}-{end}'
```
```suggestion
            iteration_key = f'{iteration_key}-{start}-{end}'
```

millerdev commented on corehq/blobs/migrate.py:
```diff
@@ -305,11 +313,35 @@ def migrate(self, doc):
+        except NotFound:
+            self.save_backup(doc)
+        else:
+            with content, NamedTemporaryFile() as buffer:
+                shutil.copyfileobj(content, buffer)
+                buffer.seek(0)
```
I'm assuming the reason for using `NamedTemporaryFile` is to work around the [`BlobStream` copy blob feature](https://github.com/dimagi/commcare-hq/blob/fc9f450ad7442e715d07a1dede61bfc887c5632c/corehq/blobs/s3db.py#L69). Would it work to use [`io.BufferedReader`](https://docs.python.org/3.7/library/io.html#io.BufferedReader) to stream directly from `content`?
```suggestion
            with content:
                buffer = BufferedReader(content)
```

snopoke replied:
I did consider this (and I'm sure I tried it in a test) but since the blobdb is reading and writing to the same key I wanted to avoid any edge cases where the write would impact the read. I'm not sure about the `S3BlobDB` but the `FilesystemBlobDB` can't read and write to the same file concurrently.

millerdev commented on corehq/blobs/migrate.py:
```diff
@@ -500,9 +539,10 @@ class gmigrator:
+MIGRATIONS = {m().slug: m for m in [
+    lambda: BackendMigrator("migrate_backend", BlobDbBackendMigrator),
+    lambda: BackendMigrator("migrate_backend_check", BlobDbBackendCheckMigrator),
+    lambda: BackendMigrator("compress_form_xml", BlobDBCompressionMigrator, type_code=CODES.form_xml),
```
I don't see why these were converted to lambdas. Can you explain that?

snopoke replied:
If they are statically created then changing any fields on the objects persist for the duration of the runtime. This isn't an issue in a management command but in test cases if there are multiple calls to the same migration then the field values from previous tests persist.

snopoke committed:
Update corehq/blobs/migrate.py

Co-Authored-By: Daniel Miller <dmiller@dimagi.com>

millerdev commented:
Ugh, the (I think mistakenly) resolved comment did not reappear when I added another comment. See https://github.com/dimagi/commcare-hq/pull/27297#discussion_r417503557

snopoke committed:
remove self.

Co-Authored-By: Daniel Miller <dmiller@dimagi.com>

snopoke commented:
> Ugh, the (I think mistakenly) resolved comment did not reappear when I added another comment. See #27297 (comment)

Thanks, I thought I'd fixed it but apparently not. Updated now: f2b0351

snopoke commented:
Build failed right at the end and unrelated to this PR
